adr/ADR-13.md

```diff
@@ -0,0 +1,290 @@
+NATS/1.0 404 No Messages
+```
+
+## Design
+
+The implementation for pull subscribe uses a combination of both no wait and
+lingering pull requests described previously.
+
+In the Go client, a simple example of the API looks as follows:
+
+```go
+sub, err := js.PullSubscribe("stream-name", "durable")
+if err != nil {
+	log.Fatal(err)
+}
+
+for {
+	msgs, err := sub.Fetch(1)
+	if err != nil {
+		log.Println("Error:", err)
+		continue
+	}
+	for _, msg := range msgs {
+		msg.Ack()
+	}
+}
+```
+
+When implementing `PullSubscribe` there are two main cases to
+consider: `Fetch(n)` and `Fetch(1)`.
+
+#### Pull(n)
+
+`Pull(n)` batch requests are implemented somewhat similarly to old style
+requests.  When making a pull request, first a no wait request is
+done to try to get the messages that may already be available as
+needed.  If there are no messages (a 404 status message error by the
+JetStream server), then a long pull request is made:
+
+```shell
+SUB _INBOX.NQ2uAIXd4GoozKOTfECtIg  1
+PUB $JS.API.CONSUMER.MSG.NEXT.bar.dur _INBOX.WvaJLnIXcj8Zf5SrxlHMTS 26
+{"batch":5,"no_wait":true}
+
+# Result of no wait request if there are no messages
+HMSG _INBOX.WvaJLnIXcj8Zf5SrxlHMTS 8  28 28
+NATS/1.0 404 No Messages
+
+# Next pull request is a long pull request with client side timeout
+PUB $JS.API.CONSUMER.MSG.NEXT.bar.dur _INBOX.NQ2uAIXd4GoozKOTfECtIg 32
+{"expires":4990000000,"batch":5}
+```
+
+As part of the request payload, the `batch` field is set to the number
+of expected messages and `expires` is set to cancel the request
+`100ms` before the client side timeout.  In the example above, the
+timeout is `5s` so in the payload `expires` is the result of `5s -
+100ms` represented in nanoseconds.
+
+After making the first request no wait request, it is also recommended
+to send an auto unsubscribe protocol discounting the message already
+received as a result of the no wait request.  In case the batch
+request was for 5 messages, the client would auto unsubscribe after
+receiving 4:
+
+```shell
+UNSUB 1 4
+```
+
+When successful, the result of a batch request will be at least one
+message being delivered to the client.  In case not all messages are
+delivered and the client times out or goes away during the pull
+request, it is recommended to unsubscribe to remove the interest of
+the awaited messages from the server, otherwise this risks the server
+sending messages to an inbox from a connected client that is no longer
+expecting messages.
+
+```shell
+SUB _INBOX.WvaJLnIXcj8Zf5SrxlHMTS  1
+PUB $JS.API.CONSUMER.MSG.NEXT.bar.dur _INBOX.WvaJLnIXcj8Zf5SrxlHMTS 26
+{"batch":5,"no_wait":true}
+HMSG _INBOX.WvaJLnIXcj8Zf5SrxlHMTS 8  28 28
+NATS/1.0 404 No Messages
+UNSUB 1 4
+PUB $JS.API.CONSUMER.MSG.NEXT.bar.dur _INBOX.WvaJLnIXcj8Zf5SrxlHMTS 32
+{"expires":4990000000,"batch":5}
+MSG hello 1 $JS.ACK.bar.dur.2.29034.29041.1626845015078897000.0 4
+ping
+# Only 1 out of 5 messages receives, so client goes away and unsubscribes.
+UNSUB 1
```

Comment:
So does that mean that the library sets a subscription for a batch and if the batch is full, we know the server does not send more, but otherwise we "timeout" the subscription if not all messages in the batch were received?
Most libs don't have a wait to timeout an async subscription, so timeout is usually with sync subscription. If that is the case, then will have to keep track of time already spent dequeueing some messages from the sub (that is, timeout will decrease overtime).

And to be clear, in the case above where 1 message was received (out of 5), this is not considered and error and that single message is returned by Fetch(5), right?

Reply:
In case of the Go client, yes only 1 message out of the 5 collected would be returned, but no `nats: timeout` error is returned for example

Reply:
a sync subscription might be fine since for batches larger than 1, a new subscription has to be made for each request, then unsubscribe once there is a timeout or the complete set of messages was delivered (just like old style requests just that instead of expecting single response expecting now many)

Reply:
I think I understand.